es5-shim.js

```diff
@@ -1921,20 +1921,93 @@
     }());
 
     if (!replaceReportsGroupsCorrectly) {
+        /**
+         * @param regexp {RegExp}
+         * @returns {Object[]} groups
+         *  @return {number} groups[].number        - The number of group.
+         *  @return {boolean} groups[].mayBeMissing - The group with quantifiers preventing its exercise, the matched text
+         *                                            for a capturing group is now undefined instead of an empty string.
+         *                                            The group without quantifiers, may be empty string.
+         *  @return {boolean} groups[].isResulting  - The groups with ?:, ?=, ?! are non-capturing groups.
+         */
+        var getExtendedGroups = function getExtendedGroups(regexp) {
+            var source = regexp.source;
+            var result = [];
+            var groupNumberCounter = 0;
+
+            for (var charIndex in source) {
+                var char = source[+charIndex];
+                var prevFirst = source[+charIndex - 1] || null;
+                var nextFirst = source[+charIndex + 1] || null;
+                var nextSecond = source[+charIndex + 2] || null;
+
+                if (char === '(') {
+                    if (prevFirst && prevFirst === '\\') {
+                        continue;
+                    }
+
+                    if (
+                        nextFirst && nextFirst === '?' && nextSecond && nextSecond === ':'
+                        || nextFirst && nextFirst === '?' && nextSecond && nextSecond === '='
+                        || nextFirst && nextFirst === '?' && nextSecond && nextSecond === '!'
+                    ) {
+                        groupNumberCounter += 1;
+                        result.push({
+                            number: groupNumberCounter,
+                            mayBeMissing: true,
+                            isResulting: false
+                        });
+                        continue;
+                    }
+
+                    result.push({
+                        number: ++groupNumberCounter,
+                        mayBeMissing: false,
+                        isResulting: true
+                    });
+                }
+
+                if (char === ')') {
+                    var groupIndex = groupNumberCounter - 1;
+
+                    if (prevFirst && prevFirst === '\\') {
+                        continue;
+                    }
+
+                    if (result[groupIndex] && !result[groupIndex].isResulting) {
+                        result.splice(groupIndex, 1);
```

Comment:
we should definitely not rely on:
 - `splice` working properly
 - `splice` being present

nor should we mutate the array like this.